### ISSUE TITLE
Implement Bachmann-Casella Newton convergence diagnostic method

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9603]
A recent paper by @bernhardbachmann and @casella ([DOI:10.1016/j.amc.2021.125991](https://doi.org/10.1016/j.amc.2021.125991), [DOI:10.48550/arXiv.1911.12433](https://doi.org/10.48550/arXiv.1911.12433)) proposes a diagnostic method to figure out which individual initial guess values are more likely to be causing the convergence failure of Newton-type iterative nonlinear solvers.

The method assigns numerical indicators to each initial guess; the larger the indicator value (particularly if above 1), the higher is the likelyhood that it should be improved to achieve convergence. This allows to rank the start values, so that one can focus on improving the ones that really can make a difference.

The method is currently being implemented in OMC by @TeusvanderStelt, and will provide textual output. If found to be effective in real-life cases, this diagnostics will eventually be integrated in the Declarative Debugger.

**Comment by TeusvanderStelt** [comment]
To be finished: 
- the step that automatically selects the non-linear equations and the non-linear dependences, i.e. according the paper: split into two sub-vectors w ∈ R q and z ∈ R m −q. 
- testing the above

**Comment by casella** [comment]
@TeusvanderStelt, for the time being you can skip the first step entirely.

Indicators of linear variables and linear equations will be computed to zero a posteriori instead of being skipped a priori, and that's it. Of course the approach will be less efficient, but since the computation is only done once at initialization duringing debugging, it is no big deal if it takes, say, 4 s instead of 0.5 s to do that. Who cares. This may become important for really large systems, e.g. power grid models with 100.000 variables and equations, of which only 5.000 are actually nonlinear, because this may mean going down from 2 minutes to 2 s, but we can tackle these cases later on.

It is really important that we can start the testing of the algorithm on small- and medium-sized systems (i.e. up to a few hundred unknowns) right now. For that, setting aside linear variables and equations is not really necessary. A more efficient implementation for large-scale systems can come later.

Thanks!

**Comment by casella** [comment]
The work is nearly complete and ready to be merged on master, currently in PR #9263.

There are still a few issues to be fixed, see
- #10114
- #10115
- #10116

**Comment by casella** [comment]
Updated test package [NewtonDiagnostics.zip](https://github.com/user-attachments/files/18132916/NewtonDiagnostics.zip)

**Comment by phannebohm** [comment]
I think we can consider this implemented. See the very nice [paper and slides from the 16th IMoC 2025](https://pretalx.com/16thmodelicafmiconference/talk/HLEDGX/).

**Comment by casella** [comment]
Not really. Unfortunately there is still a key piece missing, namely proper scaling of residuals. This is on @linuslangenkamp's TODO list

**Comment by phannebohm** [comment]
Right, sorry I missed that.
